Remove beofre all from network tests

alonSadan opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**:

Ginkgo's BeforeAll limits the ability to clean up in an AfterEach, adding a
generic global cleanup to run after each test,  and running tests in parallel.
This commit removes BeforeAll from all tests in the network test suit, and whenever
it is necessary uses BeforeEach and AfterEach.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #6123 

**Special notes for your reviewer**:
Each commit addresses different test file in the network test-suit that had before all, except for  
the vmi_multus file which was addressed by @ormergi  in #6293.

This change introduced a lot  of additional  vmi creations in the tests and it had a time cost. I roughly measured the time it takes 
to run the entire network suit with and without this change and there is a significant difference as can be shown below:   
(output of running the tests with [time](https://man7.org/linux/man-pages/man1/time.1.html) tool.

without the change:

real	123m58.789s
user	5m19.625s
sys	1m9.043s

with the change: 

real	174m19.060s
user	6m51.113s
sys	2m9.304s


This is not an accurate measure per se, and it includes some skips and failures happen on my local environment,   but it can give the reader an idea of the time cost.


**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

alonSadan commented:
Will appriciate a review from
/assign @EdDev

alonSadan commented:
I initiated the client in an inner context. Moved to the outer most context.
https://github.com/kubevirt/kubevirt/compare/dadc4cb8ab72e3620b09ced41f31512e0553c511..a4d468962fc46254e56512267a4ae077b802477c

rmohr commented on tests/network/vmi_slirp_interface.go:
```diff
@@ -49,9 +49,10 @@ var _ = SIGDescribe("[Serial]Slirp Networking", func() {
 
-	var genericVmi *v1.VirtualMachineInstance
-	var deadbeafVmi *v1.VirtualMachineInstance
+	var vmi *v1.VirtualMachineInstance
```
You should be able to remove the `Serial` tag now from these tests to allow parallel execution.

alonSadan replied:
Fixed.

rmohr commented on tests/network/vmi_networking.go:
```diff
@@ -62,8 +62,6 @@ var _ = SIGDescribe("[Serial][rfe_id:694][crit:medium][vendor:cnv-qe@redhat.com]
 	var currentConfiguration v1.KubeVirtConfiguration
 
 	var inboundVMI *v1.VirtualMachineInstance
-	var inboundVMIWithPodNetworkSet *v1.VirtualMachineInstance
```
If the tests are properly isolated now, you can remove the `Serial` tag to allow parallel execution.

rmohr commented on tests/network/vmi_networking.go:
```diff
@@ -72,12 +70,14 @@ var _ = SIGDescribe("[Serial][rfe_id:694][crit:medium][vendor:cnv-qe@redhat.com]
 		util.PanicOnError(err)
 
 		kv := util.GetCurrentKv(virtClient)
 		currentConfiguration = kv.Spec.Configuration
```
If you want to modify the configuration, you should take a snapshot after the cleanup took place.

rmohr commented on tests/network/networkpolicy.go:
```diff
@@ -40,50 +41,46 @@ var _ = SIGDescribe("[Serial][rfe_id:150][crit:high][vendor:cnv-qe@redhat.com][l
 		tests.SkipIfUseFlannel(virtClient)
 		skipNetworkPolicyRunningOnKindInfra()
-
+		tests.BeforeTestCleanup()
```
If the tests are properly isolated, you can remove the `Serial` tag now.

alonSadan replied:
Done

rmohr commented on tests/network/vmi_networking.go:
```diff
@@ -743,7 +724,7 @@ var _ = SIGDescribe("[Serial][rfe_id:694][crit:medium][vendor:cnv-qe@redhat.com]
 			)
 
-			It("[outside_connectivity]should be able to reach the outside world [IPv4]", func() {
+			FIt("[outside_connectivity]should be able to reach the outside world [IPv4]", func() {
```
You probably want `It`.

rmohr commented on tests/utils.go:
```diff
@@ -5105,3 +5106,37 @@ func VMILauncherIgnoreWarnings(virtClient kubecli.KubevirtClient) func(vmi *v1.V
+			return err
+		}
+
+		err = wait.PollImmediate(time.Second, time.Minute, func() (bool, error) {
```
You can simply write `Eventually(ThisVMI(vmi), 120).Should(BeGone())`

rmohr replied:
You can write `EventuallyWithOffset(1, ThisVMI(vmi), 120).Should(BeGone())` to see the error at the right location in the log traceback. I can understand that you want to return an error, but I would prefer to get rid of all the custom loop implementations and consolidate on the Ginkgo provided one.

EdDev replied:
`libvmi` is a helper package that may be used by different callers at different stack level calls.
Using the offset will not support such calls (from different stack levels) and we saw this is causing a mess in the error trace.

Therefore, I think it is most preferable not to assert inside helpers and let this be done at the test level. This is also much more transparent to a test reader.
Assert helpers by themselves are fine, but then we are not mixing other operations with the assert (e.g. `assertVMIExists`).

rmohr replied:
I see a few options here, but all the custom wait logic is a problem:

 1) Whenever possible, just doing the explicit `Eventually` in the test. With the advanced matchers they are pretty expressive. The waitfor helpers are not very explicit, one always has to check exactly what they are waiting for and they vary how they are implemented. We have seen issues with them too.
 2) If helpers are preferred to have e.g. a `EventuallyVMIShouldBeGone(vmi)` which can use the `Eventually inside` and maybe accompany it with a `EventuallyVMIShouldBeGoneWithOffset(1, vmi).
3) I don't mind if some deep-down helpers would return an error for some things to be easier concatenated, but in general what we do in all test helpers is just checking if we reach a goal and just fail the test without any special cleanup or countermeasures, which makes the error return pretty redundant.

EdDev replied:
> The waitfor helpers are not very explicit, one always has to check exactly what they are waiting for and they vary how they are implemented. We have seen issues with them too.

If they are implemented correctly, with established methods (in this case, `PollImmediate` from k8s), I think we can trust them.

But I am fine with helpers which do not retry internally, they just return responses and at the test level these are retried with the ginkgo tooling. I just think this is not always possible or efficient.

Option 2, to me, is just saying: `assertVMIShouldBeGoneWithOffset(1, vmi)`, which is fine, it just suffers from the inability to share this code out of this project (i.e. we couple everything to ginkgo, which is not a good strategy IMO).
Having helper packages depending on the ginkgo framework feels wrong to me in general and avoiding it will put better boundaries.
Said that, this is a good middle ground we could probably come to an agreement.

> but in general what we do in all test helpers is just checking if we reach a goal and just fail the test without any special cleanup or countermeasures, which makes the error return pretty redundant.

I am aware of this pattern. I think it is not an ideal one, suffering from things like:
- The offset issues (we do not control who calls the helpers and from what stack "distance").
- Inability to add informative info on the way "back" (e.g. add to the error relevant info if something fails).
- Inability to cleanup on a per-test or per-action. This is a framework limitation, which became a reason to work differently in kubevirt.
- Coupling to a specific framework. Which makes it hard to share the helpers with other projects or to be flexible and replace the framework if a better one comes to be.

For me, just the "offset" is a good enough reason to not embed asserts in actionable helpers. And having the version with the offset as an argument is just pure ugly IMO.

alonSadan replied:
IIUC option 2 is kind of what we have right now: 
https://github.com/kubevirt/kubevirt/blob/ae8cd8e6a66f711431c46eed31d0f6ccb3117bc6/tests/libvmi/operations.go#L38-L40

Or did I miss something?

rmohr replied:
At the end the `WaitForVmiDeletion` is not needed and directly placing the `Eventually` in the code should be good enough. I don't think the rest needs to be discussed any further.

alonSadan replied:
Done. 
What I did was to remove the helpers completely, because they didn't have much use, and the only place where we wait for the vmi deleting I returned the `Eventually` implementation.

rmohr reviewed (commented):
A few initial comments.

rmohr commented:
> without the change:
> 
> real 123m58.789s
> user 5m19.625s
> sys 1m9.043s
> 
> with the change:
> 
> real 174m19.060s
> user 6m51.113s
> sys 2m9.304s

That is not too bad. I think that there are big chunks of tests (multus, service, slirp/macvtap, ...) which can be run in parallel after the cleanup. I see 87 tests running in serial mode right now (https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/6331/pull-kubevirt-e2e-k8s-1.19-sig-network/1433054891114237952), and almost all can be run in parallel once all `BeforeAll` are gone (some don't even use it right now and could potentially just be switched to parallel).

rmohr commented:
Just for completeness, if you remove `Serial` you will not see faster time locally because by default we don't run the tests locally in parallel. You can `export KUBEVIRT_E2E_PARALLEL=true` to run tests in parallel locally.

alonSadan commented:
[rebased on master](https://github.com/kubevirt/kubevirt/compare/3512be51e1c984c7a0875e762c28d152e38125e5..92d70fdc236c381af57301421b4eada8300c7e1e)

Thank you @rmohr  for the notes, and sorry for the late reply(holidays). Will address your comments shortly.

alonSadan reviewed (commented):
Thank you @rmohr.
I addressed your comments.

> Just for completeness, if you remove Serial you will not see faster time locally because by default we don't run the tests locally in parallel. You can export KUBEVIRT_E2E_PARALLEL=true to run tests in parallel locally.

Thanks. I am running it now in parallel and will update the time measured.

rmohr commented on tests/network/networkpolicy.go:
```diff
@@ -228,9 +293,14 @@ var _ = SIGDescribe("[Serial][rfe_id:150][crit:high][vendor:cnv-qe@redhat.com][l
 			})
 			AfterEach(func() {
 				waitForNetworkPolicyDeletion(policy)
+				Expect(tests.DeleteVmiAndWaitForDeletion(virtClient, clientVMI)).To(Succeed())
```
It is not clear to me if you need to wait for the VMIs to disappear or if you just to that as a precation? In general we try to write the tests in a way that we can cope with some leftovers, since waiting for VMIs to disappear can take quite some time.

alonSadan replied:
Just a precaution. I'll go over and if it is not necessary I'll remove the wait part.

alonSadan replied:
Also, maybe if we are not waiting for the vmi to be gone, we can just rely on the global cleanup which happens after each test?

rmohr replied:
Yes. Until all `BeforeAll` are remove from the test suite, every test file does the cleanup itself with

```
BeforeEach() {
 tests.BeforeTestCleanup()
}
```

which deletes everything but does not wait, to avoid the long wait penalties.

alonSadan replied:
Tnx:)

oshoval replied:
maybe should be done in the AfterEach ? 
as this PR fixed
https://github.com/kubevirt/kubevirt/pull/6335

rmohr replied:
Yes can be done in an `AfterEach` too. A the end once all `BeforeAll` are gone we just remove all calls of `tests.BeforeTestCleanup()` and move it to one global `JustAfterEach`.

rmohr replied:
But right now you have the best guarantees if you use a `BeforeEach`.

alonSadan replied:
> maybe should be done in the AfterEach ?
as this PR fixed
#6335

IIRC currently the global cleanup with `tests.BeforeTestCleanup()` must happen before the set-up stage, because collecting debugging info on failures happens at the tear-down,  so if we clean everything at the tear-down we might miss some info.
So if I'll use an `AFfterEach` I must not rely on `tests.BeforeTestCleanup()`.

oshoval replied:
You right, but there is this new PR that make both works
https://github.com/kubevirt/kubevirt/pull/6350

Anyhow the issue there was a private case that caused seg-fault, so whatever you choose

rmohr replied:
> IIRC currently the global cleanup with `tests.BeforeTestCleanup()` must happen at the set-up stage, because collecting debugging info on failures happens at the tear-down, so if we clean everything at the tear-down we might miss some info.
> So if I'll use an `AFfterEach` I must not rely on `tests.BeforeTestCleanup()`.

Yes, but luckily that changed a few days ago: https://github.com/kubevirt/kubevirt/pull/6350/files. We collect now always properly.

alonSadan replied:
Nice. Thanks.

alonSadan commented:
> > without the change:
> > real 123m58.789s
> > user 5m19.625s
> > sys 1m9.043s
> > with the change:
> > real 174m19.060s
> > user 6m51.113s
> > sys 2m9.304s
> 
> That is not too bad. I think that there are big chunks of tests (multus, service, slirp/macvtap, ...) which can be run in parallel after the cleanup. I see 87 tests running in serial mode right now (https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/6331/pull-kubevirt-e2e-k8s-1.19-sig-network/1433054891114237952), and almost all can be run in parallel once all `BeforeAll` are gone (some don't even use it right now and could potentially just be switched to parallel).

rough time measurement when testing on  parallel: 

real	65m42.416s
user	10m16.181s
sys	2m55.991s

(slirp suit can not be run on parallel, because it alters the global kubevirt configuration, and on this run all slirp tests failed immediately, which probably affected the time of the run)

alonSadan commented:
> > (slirp suit can not be run on parallel, because it alters the global kubevirt configuration, and on this run all slirp tests failed immediately, which probably affected the time of the run)
> 
> We have one trick normally to run such tests in parallel: We enable by default all features and just add one or two negative tests which alter the configuration and only these are then run in a serial fashion. You can enable slrip by default here: https://github.com/kubevirt/kubevirt/blob/main/tests/utils.go#L742 and just disable it for the negative test.

Done. I hope it'l work now.
Also I removed the `afterEach`s deletes and relied on the `BeforeEach`s`tests.BeforeTestCleanup()` .

https://github.com/kubevirt/kubevirt/compare/9e5e09528017b9d5b4b6ea2e13c6570b90fb8802..f34d1089917f7dd5b4e6829b509967a6a2b3b2bf

rmohr commented:
looks like we have one test which needs a `[Serial]` left: https://github.com/kubevirt/kubevirt/blob/cb1725fcb58a19e4b57ad9c06bdedbabf2497994/tests/network/vmi_networking.go#L1016

EdDev commented:
The network tests have run until now serially. I be am worried about running them in parallel, mainly due to the maintenance cost.

Please do not mix the two in one PR. We need to discuss it and show the pros/cons before taking such an action.

alonSadan commented:
> > Done. I hope it'l work now.
> 
> looks like we have one test which needs a `[Serial]` left:
> 
> https://github.com/kubevirt/kubevirt/blob/cb1725fcb58a19e4b57ad9c06bdedbabf2497994/tests/network/vmi_networking.go#L1016
> 
> > Also I removed the `afterEach`s deletes and relied on the `BeforeEach`s`tests.BeforeTestCleanup()` .
> 
> Great. Using `BeforeEach` is still the best for now.

Thanks.
fixed in [this PR](https://github.com/kubevirt/kubevirt/pull/6388) for now until we decide about running in parallel.

alonSadan commented:
> The network tests have run until now serially. I am worried about running them in parallel, mainly due to the maintenance cost.
> 
> Please do not mix the two in one PR. We need to discuss it and show the pros/cons before taking such an action.

moved to :
https://github.com/kubevirt/kubevirt/pull/6388
keeping it draft and partial (not applied to all networking tests) until we decide.

EdDev reviewed (changes requested):
> tests, network, expose: remove beforeAll 

The commit message body is not fitting the change. It describes the removal of all it from all tests.

EdDev commented on tests/network/vmi_networking.go:
```diff
@@ -116,186 +116,190 @@ var _ = SIGDescribe("[Serial][rfe_id:694][crit:medium][vendor:cnv-qe@redhat.com]
-
-			// inboundVMI expects implicitly to be added to the pod network
-			inboundVMI = tests.NewRandomVMIWithEphemeralDiskAndUserdata(cd.ContainerDiskFor(cd.ContainerDiskCirros), "#!/bin/bash\necho 'hello'\n")
+		getVmiWithDefaultBridgeInterface := func() *v1.VirtualMachineInstance {
```
This and some more changes here could have been done as a per-requisite to this commit, right?
If that is the case, please provide them in focused commits because it is hard to review here multiple changes at once.

If that is not possible, let me know and I will start diving in deeper.

alonSadan replied:
It is possible. Done.

EdDev reviewed (changes requested):
>  tests, network, vmi_networking: remove BeforeAll
> 
> same reasoning as 102b17a

I suggest you summarize it in one short sentence. Ref a commit is for sure irrelevant as it may change on rebase.

Something like `BeforeAll has no teardown equivalent, therefore its usage leaks resources.`

> Also createAndWait and deleteAndWait vmi funcs are moved to the tests utils,
> beacsue those funcs are going to be used in the next commits.

So why you move them now if they are only needed in the next commit?
I suggest you extract these in a separate commit so we can focus better in the review.

EdDev reviewed (commented):
>  tests, network, slirp: remove beforeAll 

Same comment about the commit message as the prev ones.

alonSadan commented on tests/network/vmi_slirp_interface.go:
```diff
@@ -49,9 +49,10 @@ var _ = SIGDescribe("[Serial]Slirp Networking", func() {
-	var deadbeafVmi *v1.VirtualMachineInstance
+	var vmi *v1.VirtualMachineInstance
 	var container k8sv1.Container
+	var ports []v1.Port
```
moved to the relevant context.

alonSadan commented on tests/network/vmi_slirp_interface.go:
```diff
@@ -72,101 +73,112 @@ var _ = SIGDescribe("[Serial]Slirp Networking", func() {
-		genericVmi = tests.NewRandomVMIWithSlirpInterfaceEphemeralDiskAndUserdata(cd.ContainerDiskFor(cd.ContainerDiskCirros), "#!/bin/bash\necho 'hello'\n", ports)
-		deadbeafVmi = tests.NewRandomVMIWithSlirpInterfaceEphemeralDiskAndUserdata(cd.ContainerDiskFor(cd.ContainerDiskCirros), "#!/bin/bash\necho 'hello'\n", ports)
-		deadbeafVmi.Spec.Domain.Devices.Interfaces[0].MacAddress = "de:ad:00:00:be:af"
+			vmi = tests.NewRandomVMIWithSlirpInterfaceEphemeralDiskAndUserdata(cd.ContainerDiskFor(cd.ContainerDiskCirros), "#!/bin/bash\necho 'hello'\n", ports)
```
The vmi has to be created in the fixture because we are altering the object in each table entry [here](https://github.com/kubevirt/kubevirt/blob/2961171acb1c8cb61a176d9a4c20164efb9145df/tests/network/vmi_slirp_interface.go#L99).
That's why I thought making the base object  in the setup, and then alter and create in the fixture make sense.

Now, even though I changed it to use the vmi-ref instead of the mac address boolean, I still think it is better to keep it in the fixture and avoid creating two vmis in the beofreEach where we only need one in each entry.

alonSadan commented on tests/network/vmi_slirp_interface.go:
```diff
@@ -72,101 +73,112 @@ var _ = SIGDescribe("[Serial]Slirp Networking", func() {
+			setSlirpEnabled(false)
+		})
+
+		table.DescribeTable("should be able to", func(withMacAddress bool) {
```
OK. moved back for passing a ref to the vmi.

alonSadan commented on tests/network/vmi_slirp_interface.go:
```diff
@@ -177,7 +189,7 @@ var _ = SIGDescribe("[Serial]Slirp Networking", func() {
 		It("should reject VMIs with default interface slirp when it's not permitted", func() {
 			var t int64 = 0
-			vmi := v1.NewMinimalVMIWithNS(util.NamespaceTestDefault, libvmi.RandName(libvmi.DefaultVmiName))
+			vmi = v1.NewMinimalVMIWithNS(util.NamespaceTestDefault, libvmi.RandName(libvmi.DefaultVmiName))
```
no good reason. Fixed.

alonSadan reviewed (commented):
Thanks @EdDev 
I addressed your comments.

alonSadan commented:
> > tests, network, vmi_networking: remove BeforeAll
> > same reasoning as [102b17a](https://github.com/kubevirt/kubevirt/commit/102b17abe2ffa42ac2d664664daa016d3611a52d)
> 
> I suggest you summarize it in one short sentence. Ref a commit is for sure irrelevant as it may change on rebase.
> 
> Something like `BeforeAll has no teardown equivalent, therefore its usage leaks resources.`
> 

Thanks. Fixed.
> > Also createAndWait and deleteAndWait vmi funcs are moved to the tests utils,
> > beacsue those funcs are going to be used in the next commits.
> 
> So why you move them now if they are only needed in the next commit?
> I suggest you extract these in a separate commit so we can focus better in the review.

I extracted it to a different commit.

alonSadan commented:
> > tests, network, expose: remove beforeAll
> 
> The commit message body is not fitting the change. It describes the removal of all it from all tests.

Fixed

alonSadan commented:
/retest

rmohr commented:
> The network tests have run until now serially. I am worried about running them in parallel, mainly due to the maintenance cost.
> 
> Please do not mix the two in one PR. We need to discuss it and show the pros/cons before taking such an action.

Since the runtime of the lane would significantly increase I would suggest to keep it together. If individual tests would be come unreliable, they could be separately marked wit a simple `[serial]` tag in a follow-up PR.

EdDev commented:
> Since the runtime of the lane would significantly increase I would suggest to keep it together. If individual tests would be come unreliable, they could be separately marked wit a simple `[serial]` tag in a follow-up PR.

If this one is approved/lgtm, I am fine with adding a commit that adds the parallel part. But I do not want the two mixed in the main review session and I am interested in seeing that all is good without adding the parallel part in the initial party.
If there is flakiness, we will have hard time knowing what caused it.

rmohr commented:
> If this one is approved/lgtm, I am fine with adding a commit that adds the parallel part. But I do not want the two mixed in the main review session and I am interested in seeing that all is good without adding the parallel part in the initial party.
> If there is flakiness, we will have hard time knowing what caused it.

For as long as it goes in together all in-one, fine by me.

alonSadan commented:
> > If this one is approved/lgtm, I am fine with adding a commit that adds the parallel part. But I do not want the two mixed in the main review session and I am interested in seeing that all is good without adding the parallel part in the initial party.
> > If there is flakiness, we will have hard time knowing what caused it.
> 
> For as long as it goes in together all in-one, fine by me.

NP. I'll add the parallel part in another commit

alonSadan commented:
> f this one is approved/lgtm, I am fine with adding a commit that adds the parallel part. But I do not want the two mixed in the main review session and I am interested in seeing that all is good without adding the parallel part in the initial party.
> If there is flakiness, we will have hard time knowing what caused it.

Network lanes seems to pass with the current changes( before adding the parallel part). Once we'll finish the review process on the current changes i'll add the parallel commit.

EdDev commented on tests/utils.go:
```diff
@@ -5187,3 +5187,33 @@ func MountCloudInitFunc(devName string) func(*v1.VirtualMachineInstance) {
+	return vmi, nil
+}
+
+func DeleteVmi(virtClient kubecli.KubevirtClient, vmi *v1.VirtualMachineInstance) error {
+	if vmi != nil {
+		return virtClient.VirtualMachineInstance(vmi.Namespace).Delete(vmi.Name, &metav1.DeleteOptions{})
+	}
+
+	By("vmi is nil")
+	return nil
+}
```
If the VMI is nil, it should not be called. Having this helper is not really useful.

alonSadan replied:
Fixed.

EdDev commented on tests/utils.go:
```diff
@@ -5187,3 +5187,33 @@ func MountCloudInitFunc(devName string) func(*v1.VirtualMachineInstance) {
+	return nil
+}
+
+func WaitForVmiDeletion(vmi *v1.VirtualMachineInstance) {
+	if vmi != nil {
+		EventuallyWithOffset(1, ThisVMI(vmi), 120).Should(BeGone())
+	}
+}
```
Same, wrapping the nil condition is not nice.

alonSadan replied:
Fixed.

EdDev reviewed (changes requested):
Review until commit `tests, network, vmi_networking: refactor to helpers`

Splitting this to its own commit really helped see the change clearly. thanks!

alonSadan commented on tests/network/vmi_networking.go:
```diff
@@ -116,53 +116,41 @@ var _ = SIGDescribe("[Serial][rfe_id:694][crit:medium][vendor:cnv-qe@redhat.com]
-			inboundVMIWithPodNetworkSet = tests.WaitUntilVMIReady(inboundVMIWithPodNetworkSet, libnet.WithIPv6(console.LoginToCirros))
-			inboundVMIWithCustomMacAddress = tests.WaitUntilVMIReady(inboundVMIWithCustomMacAddress, libnet.WithIPv6(console.LoginToCirros))
-
-			tests.StartTCPServer(inboundVMI, testPort)
```
oops. Thanks. Fixed

alonSadan commented on tests/utils.go:
```diff
@@ -5187,3 +5187,33 @@ func MountCloudInitFunc(devName string) func(*v1.VirtualMachineInstance) {
 	}
 }
+
+func CreateAndWaitForVmi(client kubecli.KubevirtClient, vmi *v1.VirtualMachineInstance) (*v1.VirtualMachineInstance, error) {
```
This is mostly for readability, and prevent code repetitions. I don't like this long  doted line of code  for every client operation. Also many suits define their own vmi-operations helpers and can be refactored to use the global function. 
I moved it to a new file: libvmi/operations.go

alonSadan reviewed (commented):
Thanks @EdDev  for the thorough review. addressed your comments.

EdDev commented on tests/network/vmi_networking.go:
```diff
@@ -116,53 +116,41 @@ var _ = SIGDescribe("[Serial][rfe_id:694][crit:medium][vendor:cnv-qe@redhat.com]
 			for _, networkVMI := range []*v1.VirtualMachineInstance{inboundVMI, outboundVMI, inboundVMIWithPodNetworkSet, inboundVMIWithCustomMacAddress} {
-				_, err = virtClient.VirtualMachineInstance(util.NamespaceTestDefault).Create(networkVMI)
-				Expect(err).ToNot(HaveOccurred())
+				tests.CreateAndWaitForVmi(virtClient, networkVMI)
```
The creation was called, which is async. So while waiting for the first vmi to be created, the others would also in parallel get processed. When waiting for the second vmi or beyond, the waiting would be minimal if at all (as in the meantime maybe they are already up).

alonSadan commented:
> Can you please drop the get prefix? It could be replaced with new but even that is not really needed.

Done.

alonSadan commented:
[fixed two stupid mistakes](https://github.com/kubevirt/kubevirt/compare/e0a88f142a8807e010d36b84eb71caff3ba4c4ea..302e4911203c8c063356c9d205ff3a49038bb963)

EdDev commented on tests/libvmi/operations.go:
```diff
@@ -0,0 +1,31 @@
+}
+
+func WaitForVmiDeletion(vmi *v1.VirtualMachineInstance) {
+	EventuallyWithOffset(1, matcher.ThisVMI(vmi), 120).Should(matcher.BeGone())
```
Please do not use an assertion here, just return an error if the timeout is reached so the caller can handle it any way he wants. Lets keep assertions out of helpers as much as possible.

alonSadan replied:
Once we will decide on [this thread](https://github.com/kubevirt/kubevirt/pull/6331#discussion_r700997713) what should be done, I'll act accordingly. Originally I returned an error and did not assert but @rmohr asked me to assert.

EdDev commented on tests/libvmi/operations.go:
```diff
@@ -0,0 +1,31 @@
+}
+
+func DeleteVmiAndWaitForDeletion(virtClient kubecli.KubevirtClient, vmi *v1.VirtualMachineInstance) {
+	Expect(DeleteVmi(virtClient, vmi)).To(Succeed())
```
Same, lets not assert here. Just return an error and assert at the test itself.

alonSadan replied:
Once we will decide on this thread what should be done, I'll act accordingly. Originally I returned an error and did not assert but @rmohr asked me to assert.

EdDev commented on tests/libvmi/operations.go:
```diff
@@ -0,0 +1,31 @@
+	return vmi, nil
+}
+
+func DeleteVmi(virtClient kubecli.KubevirtClient, vmi *v1.VirtualMachineInstance) error {
```
Please order the functions such that the caller is before the callee. It helps reading the code from the most abstract down the details, without the need to bounce all the time.

Also, if no one is using this, keep it unexported. or, perhaps just do not implement `DeleteVmiAndWaitForDeletion`, and let the caller do it in two steps. Up to you.

EdDev commented on tests/libvmi/operations.go:
```diff
@@ -0,0 +1,31 @@
+	"kubevirt.io/kubevirt/tests/framework/matcher"
+)
+
+func CreateVmi(client kubecli.KubevirtClient, namespace string, vmi *v1.VirtualMachineInstance) (*v1.VirtualMachineInstance, error) {
```
You do not need to pass in the client, you can just fetch it here, it is a singletone anyway. It makes the API nicer IMO.
Not a must.

EdDev reviewed (commented):
Review: tests, network, vmi_networking: refactor to helpers

alonSadan reviewed (commented):
Thanks @EdDev . Addressed some of your comments. 

Your comment about asserting in the helper was discussed [here](https://github.com/kubevirt/kubevirt/pull/6331#discussion_r700997713)
Once we will decide I will act accordingly. I don't mind either way. 
Perhaps @AlonaKaplan has an input on the subject.

alonSadan commented:
/retest
There was some kind of network issue that is not related to the PR.

alonSadan commented:
I removed the helpers commit and returned to `Eventually` implementation when waiting for vmi deletion

https://github.com/kubevirt/kubevirt/compare/9c03a8a71e9179ee12c193656eb2ff22404cad39..5fcac4a28d24927daba71c4516597a9f9737392a

rmohr commented:
> I removed the helpers commit and returned to `Eventually` implementation when waiting for vmi deletion
> 
> https://github.com/kubevirt/kubevirt/compare/9c03a8a71e9179ee12c193656eb2ff22404cad39..5fcac4a28d24927daba71c4516597a9f9737392a

Sounds good. :+1:

alonSadan commented:
/cc @rmohr

alonSadan commented:
It looks like running network tests in parallel worked 
@EdDev @rmohr

rmohr commented:
/test pull-kubevirt-fossa

alonSadan commented:
/retest
